File: slaspec/instructions/pcode.py
from slaspec.const import DEFAULT_MEM


def pOP(code: str) -> str:
    return f"\n\t{code};"


def pMACRO(macro: str) -> str:
    return f"{macro}()"


def pLOCAL(var: str, size: int = 1) -> str:
    return f"local {var}:{size}"


def pVAR(var: str) -> str:
    return var


def pCOPY(var: str, val: str) -> str:
    return f"{var} = {val}"


def pPTR(addr: str, size: int = 1, mem: str = DEFAULT_MEM) -> str:
    return f"*[{mem}]:{size} {addr}"


def pRETURN(addr: str) -> str:
    return f"return [{addr}]"

File: src/slaspec/instructions/pcode.rs
use crate::slaspec::globals::DEFAULT_MEM;

pub fn p_field(field: String) -> String {
    format!("{{{field}}}")
}

pub fn p_macro(mac: String) -> String {
    format!("{mac}()")
}

pub fn p_local(var: String, size: usize) -> String {
    format!("local {var}:{size}")
}

pub fn p_copy(dst: String, src: String) -> String {
    format!("{dst} = {src}")
}

pub fn p_ptr(size: usize, addr: String) -> String {
    p_ptr_mem(DEFAULT_MEM, size, addr)
}

pub fn p_ptr_mem(mem: &str, size: usize, addr: String) -> String {
    format!("*[{mem}]:{size} {addr}")
}

pub fn p_return(addr: String) -> String {
    format!("return [{}]", addr)
}
